src/mpt.rs

```diff
@@ -0,0 +1,660 @@
+const NULL_NODE_VALUE: &[u8] = &[0_u8];
+
+/// `MptChanges` is a wrapper of changes in [Mpt] when call function close()
+#[derive(Debug, Clone)]
+pub(crate) struct MptChanges(
+    pub(crate) HashMap<Vec<u8>, Vec<u8>>,
+    pub(crate) HashSet<Vec<u8>>,
+    pub(crate) Sha256Hash,
+);
+
+/// `Mpt` is struct to maintain the Merkle Patricia Trie tree as storage struct
+///
+/// Merkle Tree: A hash tree in which each node’s hash is computed from its child nodes hashes.
+///
+/// Patricia Trie: A efficient Radix Trie (r=16), a data structure in which “keys” represent the path one has to take to reach a node
+#[derive(Debug, Clone)]
+pub(crate) struct Mpt<'a, S, V>
+where
+    S: DB + Send + Sync + Clone,
+    V: VersionProvider + Send + Sync + Clone,
+{
+    storage: KeyInstrumentedDB<'a, S, V>,
+    root_hash: Sha256Hash,
+}
+
+impl<'a, S: DB + Send + Sync + Clone, V: VersionProvider + Send + Sync + Clone> Mpt<'a, S, V> {
+    /// `new` is to create a new Trie with empty state_hash
+    /// This function should be called only once, during the first startup of fullnode
+    pub(crate) fn new(mut storage: KeyInstrumentedDB<'a, S, V>) -> Self {
+        // build an empty MPT for genesis world state
+        let mut default_root_hash = Default::default();
+        let key: Vec<u8> = RefHasher::hash(NULL_NODE_KEY).to_vec();
+        let value: Vec<u8> = NULL_NODE_VALUE.to_vec();
+        storage.put(key, value);
+        let mut genesis_mpt: Mpt<S, V> = Mpt {
+            storage: storage.clone(),
+            root_hash: default_root_hash,
+        };
+        let root_hash = match <V>::version() {
+            Version::V1 => {
+                let mut trie = TrieDBMutBuilder::<NoExtensionLayout>::new(
+                    &mut genesis_mpt,
+                    &mut default_root_hash,
+                )
+                .build();
+                trie.commit();
+                *trie.root()
+            }
+
+            Version::V2 => {
+                let mut trie = TrieDBMutBuilder::<ExtensionLayout>::new(
+                    &mut genesis_mpt,
+                    &mut default_root_hash,
+                )
+                .build();
+                trie.commit();
+                *trie.root()
+            }
+        };
+        Mpt { storage, root_hash }
+    }
+
+    /// `open` is to open the trie from give storage source and state_hash
+    pub(crate) fn open(storage: KeyInstrumentedDB<'a, S, V>, root_hash: Sha256Hash) -> Self {
+        let mpt: Mpt<S, V> = Mpt { storage, root_hash };
+        mpt
+    }
+
+    /// `get` is read and returns the value by key in a trie
+    ///
+    /// Return None if the key is not set before
+    ///
+    /// Error when state_hash does not exist or missed some trie nodes
+    pub(crate) fn get(&self, key: &[u8]) -> Result<Option<Vec<u8>>, MptError> {
+        match <V>::version() {
+            Version::V1 => {
+                let trie = TrieDBBuilder::<NoExtensionLayout>::new(self, &self.root_hash).build();
```

Comment:
Not something that can be fixed in this pull request, but I note that it feels kind of weird for Mpt to be passing self into `trie_db` like what we do throughout this file.

Something to consider for future refactors is whether we should have another type implement HashDB instead of Mpt.